### ISSUE TITLE
Investigate unrealistic leaf temperatures in enegy balance

### ISSUE THREAD

**Issue by justinmcgrath** [biocro/biocro#110]
Below is a copied comment from a different PR. When solving the energy balance equations, occasionally leaf temperatures are unrealistic. Using a different numerical method (Newton-Raphson instead of fixed point iteration) is a likely first step to address this. It happens very rarely, so it isn't high priority.

> > So I kept a constraint but set it to 30 degrees C.
> 
> A difference of 30 degrees C would be a lot. If it's producing such unrealistic values, then presumably something is wrong with the model or the way it's solved. That seems worth making a note of to revisit later, if there's not a plan to address it here.

There's definitely an issue with the solution method, but there could be something wrong with the model as well. I think switching to a Newton-Raphson solver would fix many of the unreasonable leaf temperature values, but I don't have any intention of doing that as part of this PR.

I took a look at the soybean leaf temperature differences ($\Delta T = T_{leaf} - T_{air}$) on the `main` branch and on this branch. It looks like the hard limit of $\pm 5^{\circ}$ C was being hit pretty frequently on `main`. On this branch, most of the leaf temperatures are within $5^{\circ}$ C of the air temperature, but that isn't being set by an artificial constraint. There is a small fraction of leaves hitting the hard limit of $\pm 30^{\circ}$ C. I could change the limit to 10 since it seems like the real solution is almost always within that range.

![2024-06-04 delta_T](https://github.com/biocro/biocro/assets/48919455/449eaf9f-f238-40f0-ae1a-4fb4a0ad4935)

The code I used is:

```r
library(BioCro)
library(lattice)

res <- with(soybean, {run_biocro(
    initial_values,
    parameters,
    soybean_weather[['2002']],
    direct_modules,
    differential_modules,
    ode_solver
)})

leaf_temperature_names <- grep('leaf_temperature', colnames(res), value = TRUE)

delta_T <- unlist(lapply(leaf_temperature_names, function(lt) {
    res[[lt]] - res$temp
}))

histogram(
    delta_T,
    freq = FALSE,
    breaks = seq(-30, 30, by = 0.5),
    ylim = c(0, 100),
    grid = TRUE
)
```

_Originally posted by @eloch216 in https://github.com/biocro/biocro/issues/109#issuecomment-2148468329_

**Comment by leavesandlemmas** [comment]
I have a suggestion that might work well for leaf temperature balance. As alternatives to Newton's method, consider some variant of [Brent's method](https://en.wikipedia.org/wiki/Brent%27s_method). Brent's method combines the secant method and bracketing. 

Brent's method could be a good alternative, because it uses the secant method , which essentially is Newton's method but the first derivative is estimated by a secant. The secant method has a convergence rate that is close to Newton's method. Bracketing methods search for roots inside a known interval. The [bisection method](https://en.wikipedia.org/wiki/Bisection_method) is a simple version. The bisection method is slower but safe in the sense that if the interval contains a root, it will always converge (unlike Newton's method); and it is possible to tell whether or not convergence is guaranteed from an set of brackets.  There are improved variants, and apparently the Boost library already has implementations: [example](https://www.boost.org/doc/libs/1_85_0/libs/math/doc/html/math_toolkit/root_finding_examples/cbrt_eg.html#math_toolkit.root_finding_examples.cbrt_eg.cbrt_no_derivatives).

**Comment by eloch216** [comment]
> I have a suggestion that might work well for leaf temperature balance. As alternatives to Newton's method, consider some variant of [Brent's method](https://en.wikipedia.org/wiki/Brent%27s_method). Brent's method combines the secant method and bracketing.

I think these methods are all limited to one-dimensional root finding? An issue here is that we actually have two unknowns (leaf temperature and transpiration rate). It would be nice to have an algorithm that checks for convergence on both of them. I know our current fixed-point iteration method doesn't check both, but it's something to keep in mind.

Later we would also like to simultaneously solve for four unknowns (`Ci`, `An`, `Tl`, and `E`) in a larger set of equations (FvCB + Ball-Berry + energy balance).

That being said, it would also be nice to have some general purpose code for (1) defining a set of simultaneous equations and (2) solving them with a user-selected algorithm, similar to how BioCro allows people to define a dynamical system and solve it. So Newton-Raphson, Brent's method, and fixed-point iteration could all be options that people can choose. I started working on this in the past but the code apparently only works on Windows, and I haven't been able to revisit that issue in a while. I can share the repository shortly.

**Comment by leavesandlemmas** [comment]
> I think these methods are all limited to one-dimensional root finding? An issue here is that we actually have two unknowns (leaf temperature and transpiration rate). It would be nice to have an algorithm that checks for convergence on both of them. I know our current fixed-point iteration method doesn't check both, but it's something to keep in mind.

Yes, AFAIK Brent's method doesn't have a multi-dimensional version. But the secant method does have one, called "[Broyden's Method](https://en.wikipedia.org/wiki/Broyden%27s_method)".  Broyden's method is essentially the secant method. Essentially, as the algorithm takes steps, it builds an approximation of the jacobian using secants. There's several variations on the same idea (under the umbrella of quasi-newton methods for optimization). Essentially all quasi-newton methods build an approximation of the Jacobian (called the Hessian in optimization problems) or the inverse of the Jacobian. 

Unfortunately, the guarantee for convergence doesn't generalize well to higher dimensions. The sign-checking on interval points *does* have a higher dimensional version, basically the Brouwer degree, but calculating it is not as easy anymore. That said, people have developed optimization methods with constraints like box constraints  (e.g. the quasi-newton [L-BFGS-B](https://en.wikipedia.org/wiki/Limited-memory_BFGS#L-BFGS-B)) So it might be possible to detect when the algorithm gets driven toward to impossible values. But this only means it isn't easy to check that a Newton's method (or quasi-newton methods) will converge for a general root-finding problem. 

> Later we would also like to simultaneously solve for four unknowns (`Ci`, `An`, `Tl`, and `E`) in a larger set of equations (FvCB + Ball-Berry + energy balance).

For this specific problem, I am fairly certain that it would be possible to either prove that it always has one valid solution or to check whether or not there are multiple solutions. This would make use of the specifics of the system, so it wouldn't really fit with the philosophy of separating the user-defined problem from the algorithm used to solve the problem.

**Comment by justinmcgrath** [comment]
The ability to define sets of equations and solvers separately, will greatly facilitate an evidenced-based approach that would be difficult to achieve now. Situation-specific solutions work fine for a design that separates equations from solvers, since you can write a solver specifically for that set of equations. In contrast, if the code for solver and equations is mixed together, it's impossible to compare solvers without duplicating nearly everything. If they're separate, you can write two different situation-specific solvers and try them on the same set of equations fairly easily, with little duplication.

**Comment by eloch216** [comment]
I am taking a simpler approach, and it may be helpful for getting the Broyden version set up. I should have a PR ready by the end of the day and we can discuss more there.

BTW, there is an explanation of the energy balance equations here: https://biocro.github.io/BioCro-documentation/latest/pkgdown/articles/web_only/energy_balance.html

The formatting is a bit messed up in the online version, so it may be easier to build the vignette locally instead.

**Comment by leavesandlemmas** [comment]
@eloch216 Looking forward to it. I wonder how well the double 1-d solving will work.
